src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -1,43 +1,217 @@
+        IServiceScopeFactory serviceScopeFactory,
+        GraphQLHttpMiddlewareOptions options,
+        IServiceProvider provider,
+        IHostApplicationLifetime hostApplicationLifetime)
+        : this(next, serializer, documentExecuter, serviceScopeFactory, options,
+              CreateWebSocketHandlers(serializer, documentExecuter, serviceScopeFactory, provider, hostApplicationLifetime, options))
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    protected GraphQLHttpMiddleware(
+        RequestDelegate next,
+        IGraphQLTextSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        GraphQLHttpMiddlewareOptions options,
+        IEnumerable<IWebSocketHandler<TSchema>>? webSocketHandlers = null)
+        : base(next, serializer, options, webSocketHandlers)
+    {
+        _documentExecuter = documentExecuter ?? throw new ArgumentNullException(nameof(documentExecuter));
+        _serviceScopeFactory = serviceScopeFactory ?? throw new ArgumentNullException(nameof(serviceScopeFactory));
+        var getRule = new HttpGetValidationRule();
+        _getValidationRules = DocumentValidator.CoreRules.Append(getRule).ToArray();
+        _getCachedDocumentValidationRules = new[] { getRule };
+        var postRule = new HttpPostValidationRule();
+        _postValidationRules = DocumentValidator.CoreRules.Append(postRule).ToArray();
+        _postCachedDocumentValidationRules = new[] { postRule };
+    }
+
+    private static IEnumerable<IWebSocketHandler<TSchema>> CreateWebSocketHandlers(
+        IGraphQLSerializer serializer,
+        IDocumentExecuter<TSchema> documentExecuter,
+        IServiceScopeFactory serviceScopeFactory,
+        IServiceProvider provider,
+        IHostApplicationLifetime hostApplicationLifetime,
+        GraphQLHttpMiddlewareOptions options)
+    {
+        if (hostApplicationLifetime == null)
+            throw new ArgumentNullException(nameof(hostApplicationLifetime));
+        if (provider == null)
+            throw new ArgumentNullException(nameof(provider));
+
+        return new IWebSocketHandler<TSchema>[] {
+            new WebSocketHandler<TSchema>(serializer, documentExecuter, serviceScopeFactory, options,
+                hostApplicationLifetime, provider.GetService<IWebSocketAuthenticationService>()),
+        };
+    }
+
+    ******************************/
+
+    /// <inheritdoc/>
+    protected override async Task<ExecutionResult> ExecuteScopedRequestAsync(HttpContext context, GraphQLRequest request, IDictionary<string, object?> userContext)
+    {
+        using var scope = _serviceScopeFactory.CreateScope();
+        try
+        {
+            return await ExecuteRequestAsync(context, request, scope.ServiceProvider, userContext);
+        }
+        finally
+        {
+            if (scope is IAsyncDisposable ad)
+                await ad.DisposeAsync();
+        }
+    }
```

Comment:
```suggestion
    protected override async Task<ExecutionResult> ExecuteScopedRequestAsync(HttpContext context, GraphQLRequest? request, IDictionary<string, object?> userContext)
    {
        var scope = _serviceScopeFactory.CreateScope();
        if (scope is IAsyncDisposable ad)
        {
            await using (ad.ConfigureAwait(false))
                return await ExecuteRequestAsync(context, request, scope.ServiceProvider, userContext);
        }
        else
        {
            using (scope)
                return await ExecuteRequestAsync(context, request, scope.ServiceProvider, userContext);
        }
    }
```

So in case when scope actually implements `IAsyncDisposable` it will be disposed twice + I added `ConfigureAwait(false)`.

Reply:
My implementation matches Microsoft's reference implementation; I will leave it as-is.

Reply:
https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceProviderServiceExtensions.cs#L111-L113

https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/AsyncServiceScope.cs

Reply:
Actually I am wrong - MS's implementation does not dispose twice.  I will accept your changes.